src/main/groovy/com/deploygate/gradle/plugins/tasks/LoginTask.groovy

```diff
@@ -16,141 +18,151 @@ import java.util.concurrent.TimeUnit
-                throw new RuntimeException('Failed to retrieve DeployGate credentials. Please try again or specify it in your build.gradle script.')
-
-        project.deploygate.appOwnerName =
-                [project.deploygate.appOwnerName, System.getenv(DeployGatePlugin.ENV_NAME_APP_OWNER_NAME), System.getenv(DeployGatePlugin.ENV_NAME_APP_OWNER_NAME_V1), localCredential.name].find {
-                    it != null
-                }
-        project.deploygate.apiToken =
-                [project.deploygate.apiToken, System.getenv(DeployGatePlugin.ENV_NAME_API_TOKEN), localCredential.token].find {
-                    it != null
-                }
-    }
+        if (!(deployGateExtension.appOwnerName && deployGateExtension.apiToken)) {
+            if (!setupCredential()) {
+                throw new RuntimeException('We could not retrieve DeployGate credentials. Please make sure you have configured app owner name and api token or any browser application is available to launch the authentication flow.')
+            }
 
-    boolean hasCredential() {
-        hasCreadentialInScript() || hasSavedCredential() || hasCredentialInEnv()
-    }
+            println "Welcome ${credentialStore.name}!"
 
-    boolean hasCredentialInEnv() {
-        [System.getenv(DeployGatePlugin.ENV_NAME_APP_OWNER_NAME), System.getenv(DeployGatePlugin.ENV_NAME_APP_OWNER_NAME_V1)].any() &&
-                [System.getenv(DeployGatePlugin.ENV_NAME_API_TOKEN)].any()
-    }
+            logger.info("The authentication has succeeded. The application owner name is ${credentialStore.name}.")
 
-    // From Gradle 7.0, only one method to get boolean property value is allowed
-    // whereas Groovy generates both isSaved() and getSaved(), Gradle 7.0 considers 2 getters exist for saved property.
-    // To suppress getSaved() we explicitly define isSaved().
-    private boolean isSaved() {
-        saved
-    }
+            // We can set the values iff it's not set yet because of the idempotency.
 
-    private boolean hasSavedCredential() {
-        localCredential.name && localCredential.token
-    }
+            if (!deployGateExtension.appOwnerName) {
+                deployGateExtension.setAppOwnerName(credentialStore.name)
+            }
 
-    private boolean hasCreadentialInScript() {
-        project.deploygate.appOwnerName && project.deploygate.apiToken
+            if (!deployGateExtension.apiToken) {
+                deployGateExtension.setApiToken(credentialStore.token)
+            }
+        }
     }
 
-    def setupCredential() {
-        saved = false
+    /**
+     * Launch the authentication flow and fetch the credentials if possible
+     * @return true if the credentials are persisted, otherwise false.
+     */
+    private boolean setupCredential() {
         if (BrowserUtils.hasBrowser()) {
             setupBrowser()
         } else {
             setupTerminal()
         }
+
+        if (!onetimeKey || !retrieveCredentialFromKey(onetimeKey)) {
```

Comment:
The previous code relayed `boolean` heavily, so the current code just checks `onetimeKey` presence.